File: core/src/main/java/com/dmi/perfectreader/book/data/BookData.java
package com.dmi.perfectreader.book.data;

import com.dmi.perfectreader.book.position.Position;

import java.io.IOException;

/**
 * Основное правило - любой позиции должна всегда соответствовать единственная страница, независимо от того, как эта позиция была достигнута
 */
public interface BookData {
    ItemReader itemReader(Position position) throws IOException;

    /**
     * Читать элементы до position, которые влияют на элементы, находящиеся на и после position
     * (например, такими элементами могут быть символ параграфа \n, маркер изменения стиля, начало новой строки в таблице)
     */
    ItemReader influenceItemReader(Position position) throws IOException;
}

File: app/src/main/kotlin/com/dmi/perfectreader/fragment/book/BookData.kt
package com.dmi.perfectreader.fragment.book

import android.net.Uri
import com.dmi.perfectreader.app.dataAccessAsync
import com.dmi.perfectreader.data.UserData
import com.dmi.perfectreader.fragment.book.content.Content
import com.dmi.perfectreader.fragment.book.location.Location

class BookData(
        private val userData: UserData,
        private val uri: Uri,
        val content: Content
) {
    var location = userData.loadBookLocation(uri) ?: Location(0.0)
        set(value) {
            dataAccessAsync {
                userData.saveBookLocation(uri, value)
            }
            field = value
        }
}